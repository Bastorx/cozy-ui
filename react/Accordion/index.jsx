import React, { Component } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import Icon from '../Icon'
import styles from './styles.styl'

class AccordionItem extends Component {
  constructor(props) {
    super(props)
    this.state = { selected: false }
  }

  render() {
    const { children, label, className } = this.props

    const { selected } = this.state

    const removeSpaceCharacters = str => str.replace(/\s+/g, '')

    const bodyName = removeSpaceCharacters(label)

    return (
      <div className={cx(styles['c-accordion-item'], className)}>
        <div
          className={cx(
            styles['c-accordion-title'],
            {
              [styles['is-active']]: selected
            },
            className
          )}
          aria-selected={selected}
          aria-controls={bodyName}
          role="tab"
          tabIndex="0"
          onClick={() => this.toggleSelection()}
        >
          <Icon icon="forward" />
          {label}
        </div>
        <div
          className={cx(
            styles['c-accordion-body'],
            {
              [styles['is-active']]: selected
            },
            className
          )}
          aria-hidden={!selected}
          aria-labelledby={bodyName}
          role="tabpanel"
        >
          {children}
        </div>
      </div>
    )
  }

  toggleSelection = () => {
    this.setState(previousState => ({ selected: !previousState.selected }))
  }
}

class Accordion extends Component {
  render() {
    const { children, className } = this.props
    return (
      <div className={cx(styles['c-accordion'], className)}>{children}</div>
    )
  }
}

Accordion.propTypes = {
  className: PropTypes.string
}

AccordionItem.propTypes = {
  label: PropTypes.string.isRequired,
  className: PropTypes.string
}

Object.assign(Accordion, { AccordionItem })

export default Accordion

export { AccordionItem }
