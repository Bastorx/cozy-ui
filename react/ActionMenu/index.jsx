import React, { Component } from 'react'
import ReactDOM from 'react-dom'
import Hammer from 'hammerjs'
import Menu from '../Menu'
import styles from './styles.styl'
import Overlay from '../Overlay'

/**
 * Use an ActionMenu to show the user possible actions to the user
 * at the bottom of the screen in a modal way.
 *
 * - Animates in/out
 * - Possible to use Escape to close (thanks to `<Overlay />`)
 * - Possible to click outside to close (thanks to `<Overlay />`)
 * - Reacts to gestures with HammerJS
 */
class ActionMenu extends Component {
  componentDidMount () {
    this.initialAppear()
    this.attachEvents()
  }

  componentWillUnmount () {
    this.gesturesHandler.destroy()
  }

  initialAppear () {
    this.turnTransitionsOff()
    this.applyTransformation(1)
    requestAnimationFrame(() => {
      this.turnTransitionsOn()
      this.applyTransformation(0)
    })
  }

  turnTransitionsOn () {
    this.menuNode.classList.add(styles['with-transition'])
  }

  turnTransitionsOff (cb) {
    this.menuNode.classList.add(styles['with-transition'])
  }

  attachEvents () {
    this.gesturesHandler = new Hammer.Manager(this.menuNode, {
      recognizers: [[Hammer.Pan, { direction: Hammer.DIRECTION_VERTICAL }]]
    })

    // to be completely accurate, `maximumGestureDelta` should be the difference between the top of the menu and the
    // bottom of the page; but using the height is much easier to compute and accurate enough.
    const maximumGestureDistance = this.menuNode.getBoundingClientRect().height
    // between 0 and 1, how far down the gesture must be to be considered complete upon release
    const minimumCloseDistance = 0.6
    // a gesture faster than this will dismiss the menu, regardless of distance traveled
    const minimumCloseVelocity = 0.6

    let currentGestureProgress = null

    this.gesturesHandler.on('panstart', e => {
      this.turnTransitionsOff()
      currentGestureProgress = 0
    })

    this.gesturesHandler.on('pan', e => {
      currentGestureProgress = e.deltaY / maximumGestureDistance
      this.applyTransformation(currentGestureProgress)
    })

    this.gesturesHandler.on('panend', e => {
      this.turnTransitionsOn()
      // Dismiss the menu if the swipe pan was bigger than the treshold,
      // or if it was a fast, downward gesture
      let shouldDismiss = e.deltaY / maximumGestureDistance >= minimumCloseDistance ||
                          (e.deltaY > 0 && e.velocity >= minimumCloseVelocity)

      if (shouldDismiss) {
        if (currentGestureProgress >= 1) {
          // Menu is already hidden, close it right away
          this.close()
        } else {
          this.animateClose()
        }
      } else {
        this.applyTransformation(0)
      }
    })
  }

  /**
   * Applies a css trasnform to the element, based on the progress of the gesture
   * @param  {Float} progress - Amount of progress between 0 and 1
   */
  applyTransformation (progress) {
    // constrain between 0 and 1
    progress = Math.min(1, Math.max(0, progress))
    this.menuNode.style.transform = 'translateY(' + (progress * 100) + '%)'
  }

  animateClose = () => {
    // we need to transition the menu to the bottom before dismissing it
    this.menuNode.addEventListener('transitionend', this.close, false)
    this.applyTransformation(1)
  }

  close = () => {
    // Remove the event handler so subsequent transitions don't trigger dismissals
    this.menuNode.removeEventListener('transitionend', this.close)
    this.applyTransformation(0)
    this.props.onClose()
  }

  handleMenuRef = menu => {
    this.menuNode = ReactDOM.findDOMNode(menu)
  }

  render () {
    const props = this.props
    return (
      <div className={styles.ActionMenu}>
        <Overlay onClick={this.animateClose} onEscape={this.animateClose}>
          <Menu className={styles['c-actionmenu']} {...props} ref={this.handleMenuRef} />
        </Overlay>
      </div>
    )
  }
}

export default ActionMenu
