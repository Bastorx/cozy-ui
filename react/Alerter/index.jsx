import React, { Component } from 'react'
import classNames from 'classnames'

import styles from './index.styl'

const AUTOCLOSE_DELAY = 2000

class Alerter extends Component {

  constructor (props) {
    super()
    this.state = {
      hidden: true
    }
    this.shouldAutoClose = !props.buttonText
    this.closeTimer = null
  }

  componentDidMount () {
    if (this.shouldAutoClose) {
      this.closeTimer = setTimeout(() => {
        this.setState({ hidden: true })
      }, AUTOCLOSE_DELAY)
    }
    // Delay to trigger CSS transition after the first render.
    // Totally open for a better way to achieve this.
    setTimeout(() => {
      this.setState({ hidden: false })
    }, 20)
  }

  componentWillUnmount () {
    this.setState({ hidden: false })
    if (this.closeTimer) {
      clearTimeout(this.closeTimer)
    }
  }

  render () {
    const { message, type, buttonText, buttonAction } = this.props
    const { hidden } = this.state

    return (<div className={styles['coz-alerter']}>
      <div
        className={classNames(
          styles['coz-alert'],
          styles[`coz-alert--${type}`],
          hidden ? styles['coz-alert--hidden'] : ''
        )}
      >
        <p>{message}</p>
        {buttonText &&
          <button onClick={buttonAction} className={classNames('coz-btn', `coz-btn--alert-${type}`)}>
            {buttonText}
          </button>
        }
      </div>
    </div>)
  }
}

export default Alerter
