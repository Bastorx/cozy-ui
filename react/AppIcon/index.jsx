import React, { Component } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import styles from './styles.styl'

import Icon from '../Icon'

import appDefaultIcon from '../../assets/icons/ui/cube.svg'
import { getPreloaded, preload } from './Preloader'

const DONE = 'done'
const ERRORED = 'errored'
const FETCHING = 'fetching'

export class AppIcon extends Component {
  constructor(props, context) {
    super(props, context)
    const { app, domain, secure } = props
    const preloaded = getPreloaded(app, domain, secure)
    this.state = {
      error: null,
      icon: preloaded,
      status: preloaded ? DONE : FETCHING
    }
  }

  componentDidMount() {
    this.load()
  }

  async load() {
    const { app, domain, fetchIcon, onReady, secure } = this.props
    const loadFn = fetchIcon || preload
    let loadedUrl
    let loadError
    try {
      loadedUrl = await loadFn(app, domain, secure)
    } catch (error) {
      loadError = error
    }

    this.setState({
      error: loadError,
      icon: loadedUrl,
      status: loadError ? ERRORED : DONE
    })

    if (typeof onReady === 'function') {
      onReady()
    }
  }

  render() {
    const { alt, className } = this.props
    const { icon, status } = this.state
    switch (status) {
      case FETCHING:
        return (
          <div
            className={cx(
              styles['c-loading-placeholder'],
              styles['c-app-icon'],
              className
            )}
          />
        )
      case DONE:
        return (
          <img
            alt={alt}
            className={cx(styles['c-app-icon'], className)}
            src={icon}
          />
        )
      case ERRORED:
      default:
        return (
          <Icon
            className={cx(
              styles['c-app-icon'],
              styles['c-app-icon-default'],
              className
            )}
            height="100%"
            icon={appDefaultIcon}
            width="100%"
          />
        )
    }
  }
}

AppIcon.propTypes = {
  alt: PropTypes.string,
  app: PropTypes.object,
  className: PropTypes.string,
  domain: PropTypes.string,
  fetchIcon: PropTypes.func,
  onReady: PropTypes.func,
  secure: PropTypes.bool
}

export default AppIcon
