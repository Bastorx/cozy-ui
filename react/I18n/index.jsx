/**
 * Provides an I18n helper using a Higher Order Component.
 */

'use strict'

import React, { Component, useContext } from 'react'
import PropTypes from 'prop-types'

import { initTranslation } from './translation'
import { initFormat } from './format'

export const DEFAULT_LANG = 'en'

const I18nContext = React.createContext()

// Provider root component
export class I18n extends Component {
  constructor(props) {
    super(props)
    this.init(this.props)
  }

  init(props) {
    const { polyglot, lang, dictRequire, context, defaultLang } = props

    this.translator =
      polyglot || initTranslation(lang, dictRequire, context, defaultLang)
    this.format = initFormat(lang, defaultLang)
    this.t = this.translator.t.bind(this.translator)
    this.contextValue = this.getContextValue()
  }

  getContextValue() {
    return {
      t: this.t,
      f: this.format,
      lang: this.props.lang
    }
  }

  // for preact
  // eslint-disable-next-line react/no-deprecated
  componentWillReceiveProps(nextProps) {
    this.UNSAFE_componentWillReceiveProps(nextProps)
  }

  UNSAFE_componentWillReceiveProps = nextProps => {
    if (nextProps.lang !== this.props.lang) {
      this.init(nextProps)
    }
  }

  render() {
    return (
      <I18nContext.Provider value={this.contextValue}>
        {this.props.children}
      </I18nContext.Provider>
    )
  }
}

I18n.propTypes = {
  lang: PropTypes.string.isRequired, // current language.
  polyglot: PropTypes.object, // A polyglot instance.
  dictRequire: PropTypes.func, // A callback to load locales.
  context: PropTypes.string, // current context.
  defaultLang: PropTypes.string // default language. By default is 'en'
}

I18n.defaultProps = {
  defaultLang: DEFAULT_LANG
}

I18n.childContextTypes = {
  t: PropTypes.func,
  f: PropTypes.func,
  lang: PropTypes.string
}

// higher order decorator for components that need `t` and/or `f`
export const translate = () => WrappedComponent => {
  const Wrapper = props => {
    const i18nContext = useContext(I18nContext)
    return (
      <WrappedComponent
        {...props}
        t={i18nContext && i18nContext.t}
        f={i18nContext && i18nContext.f}
        lang={i18nContext && i18nContext.lang}
      />
    )
  }
  Wrapper.displayName = `withI18n(${WrappedComponent.displayName ||
    WrappedComponent.name})`
  Wrapper.propTypes = {
    //!TODO Remove this check after fixing https://github.com/cozy/cozy-drive/issues/1848
    ...(WrappedComponent ? WrappedComponent.propTypes : undefined)
  }
  return Wrapper
}

export const useI18n = () => useContext(I18nContext)

export { initTranslation, extend } from './translation'

export default I18n
