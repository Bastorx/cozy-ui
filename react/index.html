<!doctype html>
<html>
	<head>
		<meta charset="utf-8">
		<title>Cozy Ui Style Guide</title>
		<meta name="viewport" content="width=device-width, initial-scale=1.0">
	<link href="main.css" rel="stylesheet"></head>
	<body>
		<div id="app"></div>
	<script type="text/javascript" src="build/bundle.180d09cd.js"></script></body>
</html>
