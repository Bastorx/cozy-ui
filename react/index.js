export {default as Alerter} from './Alerter'
export {default as Button} from './Button'
export {default as I18n, translate} from './I18n'
export {default as Icon} from './Icon'
export {default as Sidebar} from './Sidebar'
export {default as Modal, ModalContent, ModalSection, ModalButtons} from './Modal'
export {default as SelectionBar} from './SelectionBar'
export {default as Spinner} from './Spinner'
export {Tabs, TabPanels, TabPanel, TabList, Tab} from './Tabs'
export {default as Toggle} from './Toggle'
export {default as withBreakpoints} from './helpers/withBreakpoints'
