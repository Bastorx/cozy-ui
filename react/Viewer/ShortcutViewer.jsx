import React from 'react'
import { ButtonLink } from 'cozy-ui/transpiled/react/Button'
import { useI18n } from '../I18n'
import { useClient, useFetchShortcut } from 'cozy-client'
import get from 'lodash/get'
import NoViewer from './NoViewer'
import { FileDoctype } from '../proptypes'

const ShortcutViewer = ({ file }) => {
  const client = useClient()
  const { t } = useI18n()
  const { shortcutInfos } = useFetchShortcut(client, file.id)
  let url = ''
  if (shortcutInfos) {
    url = new URL(shortcutInfos.data.attributes.url)
  }
  return (
    <NoViewer
      file={file}
      renderFallbackExtraContent={() => (
        <ButtonLink
          label={`${t('Viewer.goto', { url: get(url, 'origin', '') })}`}
          icon="openwith"
          href={`${get(url, 'origin', '')}`}
          target="_blank"
        />
      )}
    />
  )
}

ShortcutViewer.propTypes = {
  file: FileDoctype.isRequired
}

export default ShortcutViewer
