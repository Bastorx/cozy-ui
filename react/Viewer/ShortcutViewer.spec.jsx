import React from 'react'

import CozyClient, { CozyProvider } from 'cozy-client'
import I18n from '../I18n'
import { render, wait } from '@testing-library/react'

import ShortcutViewer from './ShortcutViewer'
import { locales } from './index'
const setup = () => {
  const client = new CozyClient({})
  return (
    <CozyProvider client={client}>
      <I18n lang="en" dictRequire={() => locales['en']}>
        <ShortcutViewer file={{ id: '1' }} />
      </I18n>
    </CozyProvider>
  )
}

describe('Shortcutviewer', () => {
  it('renders the component', async () => {
    const { container } = render(setup())
    expect(container).toMatchSnapshot()
    await wait()
  })
})
