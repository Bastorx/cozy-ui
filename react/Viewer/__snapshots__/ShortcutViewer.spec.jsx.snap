// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shortcutviewer renders the component 1`] = `
<div>
  <div
    class="styles__viewer-noviewer___3heA6"
  >
    <svg
      class="styles__icon___23x3R"
      height="140"
      width="160"
    >
      <use
        xlink:href="#file-type-files"
      />
    </svg>
    <p
      class="styles__viewer-filename___36bWW"
    />
    <a
      class="styles__c-btn___-2Vnj"
      href=""
      target="_blank"
    >
      <span>
        <svg
          aria-hidden="true"
          class="styles__icon___23x3R"
          focusable="false"
          height="16"
          width="16"
        >
          <use
            xlink:href="#openwith"
          />
        </svg>
        <span>
          Go to 
        </span>
      </span>
    </a>
  </div>
</div>
`;
